include/graybat/communicationPolicy/socket/Base.hpp

```diff
@@ -677,41 +680,58 @@ namespace graybat {
                                                        graybat::communicationPolicy::Tag<T_CommunicationPolicy> const tag,
                                                        T_Recv& recvData)
             -> void {
-
-                using Message   = graybat::communicationPolicy::socket::Message<T_CommunicationPolicy>;
-
-                Message message(std::move(inBox.waitDequeue(msgType, context.getID(), srcVAddr, tag)));
-                memcpy (static_cast<void*>(recvData.data()),
-                        static_cast<std::int8_t*>(message.getData()),
-                        sizeof(typename T_Recv::value_type) * recvData.size());
-
+				recvImpl(msgType, context, srcVAddr, tag, recvData.data(), recvData.size());
             }
 
-
             template <typename T_CommunicationPolicy>
             template <typename T_Recv>
             auto Base<T_CommunicationPolicy>::recvImpl(graybat::communicationPolicy::Context<T_CommunicationPolicy> const context,
                                                        T_Recv& recvData)
             -> graybat::communicationPolicy::Event<T_CommunicationPolicy> {
 
+				using CommunicationPolicy = T_CommunicationPolicy;
+				using Event   = graybat::communicationPolicy::Event<CommunicationPolicy>;
+				using Message = graybat::communicationPolicy::socket::Message<CommunicationPolicy>;
+
+				std::tuple<MsgType, ContextID, VAddr, Tag> keys;
+				VAddr destVAddr;
+				Tag tag;
+
+				Message message(std::move(inBox.waitDequeue(keys, MsgType::PEER, context.getID())));
+				destVAddr = std::get<2>(keys);
+				tag = std::get<3>(keys);
+
+				memcpy (static_cast<void*>(recvData.data()),
+						static_cast<std::int8_t*>(message.getData()),
+						sizeof(typename T_Recv::value_type) * recvData.size());
+
+				return Event(getMsgID(), context, destVAddr, tag, *(static_cast<CommunicationPolicy*>(this)));
+
+			}
+
+            template <typename T_CommunicationPolicy>
+            template <typename T_Recv>
+            auto Base<T_CommunicationPolicy>::recvImpl(graybat::communicationPolicy::MsgType<T_CommunicationPolicy> const msgType,
+                                                       graybat::communicationPolicy::Context<T_CommunicationPolicy> const context,
+                                                       graybat::communicationPolicy::VAddr<T_CommunicationPolicy> const srcVAddr,
+                                                       graybat::communicationPolicy::Tag<T_CommunicationPolicy> const tag,
+													   T_Recv* recvData,
+													   const size_t size)
+			-> void
+			{
+
```

Comment:
I don't understand, where you would want to go with such a policy and how to sensefully isolate such a policy.
Can you make a little draft, how you want it to be, than i can make the changes.